File: HW2/HW2.m
x = linspace(-5, 5, 100);
y = linspace(-5, 5, 100);
[xx, yy] = meshgrid(x, y);
f1 = @(a, i, j) (2 * a) ./ sqrt((a-i).^2+j.^2);
Z1 = zeros(100, 100);
for i = 1:100
    for j = 1:100
        b = xx(i,j);
        c = yy(i,j);
        int = integral(@(a) f1(a, b, c), 0, 1);
        if int > 2.5
            Z1(i,j) = 2.5;
        else
            Z1(i,j) = int;
        end
    end
end

figure
contour(xx, yy, Z1, 20)
title("Electric field and potential for line charge (units of k)")
colorbar()
xlabel('x')
ylabel('y')
[U1, V1] = gradient(Z1);
hold on
quiver(xx, yy, U1, V1);
hold off

f2x = @(a, i, j) (a.^2) ./ sqrt((a-i).^2+j.^2);
f2y = @(a, i, j) (a) ./ sqrt(i.^2+(a-j).^2);
Z2 = zeros(100, 100);
for i = 1:100
    for j = 1:100
        b = xx(i,j);
        c = yy(i,j);
        int = integral(@(a) f2x(a, b, c), 0, 1);
        int = int + integral(@(a) f2y(a, b, c), 1, 2);
        if int > 2.5
            Z2(i,j) = 2.5;
        else
            Z2(i,j) = int;
        end
    end
end

figure
contour(xx, yy, Z2, 20);
title("Electric field and potential for L charge (units of k)")
colorbar()
xlabel('x')
ylabel('y')
[U2, V2] = gradient(Z2);
hold on
quiver(xx, yy, U2, V2);
hold off

f3 = @(a, b, i, j) a ./ sqrt((a-i).^2+(b-j).^2);
Z3 = zeros(100, 100);
ymin = @(x) -sqrt(4-x.^2);
ymax = @(x) sqrt(4-x.^2);
for i = 1:100
    for j = 1:100
        c = xx(i,j);
        d = yy(i,j);
        int = integral2(@(a, b) f3(a, b, c, d), -2, 2, ymin, ymax);
        if int > 5
            Z3(i,j) = 5;
        elseif int < -5
            Z3(i,j) = -5;
        else
            Z3(i,j) = int;
        end
    end
end

figure
contour(xx, yy, Z3, 20);
title("Electric field and potential for disk charge (units of k)")
colorbar()
xlabel('x')
ylabel('y')
[U3, V3] = gradient(Z3);
hold on
quiver(xx, yy, U3, V3);
hold off

File: HW2/HW2.py
import math
import numpy as np
import numpy.linalg as linalg
import scipy.integrate as sci
import matplotlib.pyplot as plt
from myint import *


def epotentialx(rho, a, b, x, y):
    """
    Calculates the electric potential in 2 dimensions for a 1-dimensional charge distribution

    :param rho: charge distribution as a function over x
    :param a, b: limits for integration
    :param x, y: where the potential will be calculated
    :return: V, but cut off at some maximum to avoid the singularity right at the charge itself
    """

    dV = lambda xr: rho(xr) / linalg.norm([(x - xr), y])
    V = simpsonrule(f=dV, a=a, b=b, N=100)

    return np.where(V <= 2.5, V, 2.5)


def epotentialy(rho, a, b, x, y):
    """
    Calculates the electric potential in 2 dimensions for a 1-dimensional charge distribution

    :param rho: charge distribution as a function over y
    :param a, b: limits for integration
    :param x, y: where the potential will be calculated
    :return: V, but cut off at some maximum to avoid the singularity right at the charge itself
    """

    dV = lambda yr: rho(yr) / linalg.norm([x, (y-yr)])
    V = simpsonrule(f=dV, a=a, b=b, N=100)

    return np.where(V <= 2.5, V, 2.5)


def epotentialxy(rho, ax, bx, ay, by, x, y):
    """
    Calculates the electric potential for a 2-dimensional charge distribution

    :param rho: charge distribution as a function over y
    :param ax, bx, ay, by: limits for integration
    :param x, y: where the potential will be calculated
    :return: V, but cut off at some maximum to avoid the singularity right at the charge itself
    """

    dV = lambda xr, yr: rho(xr, yr) / linalg.norm([(x - xr), (y - yr)])
    V, Verr = sci.dblquad(dV, a=ax, b=bx, gfun=ay, hfun=by)
    return V


rho1 = lambda x: 2 * x
x = np.arange(-5.0, 5.0, 0.05)
y = np.arange(-5.0, 5.0, 0.05)
Xm, Ym = np.meshgrid(x, y, sparse=True)
Z = epotentialx(rho1, 0, 1, Xm, Ym)
im = plt.imshow(Z, cmap='Reds', extent=(-5, 5, -5, 5))
plt.plot([0, 1], [0, 0], c='w')
plt.colorbar(im)
plt.title("Electric potential for line charge (units of k)")
plt.xlabel("x")
plt.ylabel("y")
plt.show()
plt.close()

U, V = np.gradient(Z)
strm = plt.streamplot(x, y, V, U, density=2, color=V, cmap='Reds')
plt.colorbar(strm.lines)
plt.title("Electric field for line charge (units of k)")
plt.xlabel("x")
plt.ylabel("y")
plt.show()
plt.close()


# Second charge distribution
rho2x = lambda x: x**2
rho2y = lambda y: y
x = np.arange(-5.0, 5.0, 0.05)
y = np.arange(-5.0, 5.0, 0.05)
Xm, Ym = np.meshgrid(x, y, sparse=True)
Z = epotentialx(rho2x, 0, 1, Xm, Ym)
Z = Z + epotentialy(rho2y, 1, 2, Xm, Ym)
im = plt.imshow(Z, cmap='Reds', extent=(-5, 5, 5, -5))
plt.plot([0, 1], [0, 0], c='w')
plt.plot([0, 0], [1, 2], c='w')
plt.colorbar(im)
plt.title("Electric potential for L charge (units of k)")
plt.xlabel("x")
plt.ylabel("y")
plt.show()
plt.close()

U, V = np.gradient(Z)
strm = plt.streamplot(x, y, V, U, density=2, color=V, cmap='Reds')
plt.colorbar(strm.lines)
plt.title("Electric field for L charge (units of k)")
plt.xlabel("x")
plt.ylabel("y")
plt.show()
plt.close()


# Third charge distribution
rho3 = lambda y, x: x
x = np.arange(-5.0, 5.0, 0.5)
y = np.arange(-5.0, 5.0, 0.5)
Z = np.empty((len(x), len(y)))
for i in range(len(x)):
    for j in range(len(y)):
        if x[i]**2 + y[j]**2 > 4.5:
            Z[i][j] = epotentialxy(rho3, ax=-2, bx=2, ay=lambda xr: -math.sqrt(4-xr**2), by=lambda xr: math.sqrt(4-xr**2), x=x[i], y=y[j])
        else:
            Z[i][j] = y[j]
im = plt.imshow(Z, cmap='Reds', extent=(-5, 5, 5, -5))
plt.colorbar(im)
plt.title("Electric potential for disk charge (units of k)")
plt.xlabel("x")
plt.ylabel("y")
plt.show()
plt.close()

U, V = np.gradient(Z)
strm = plt.streamplot(x, y, V, U, density=2, color=V, cmap='Reds')
plt.colorbar(strm.lines)
plt.title("Electric field for disk charge (units of k)")
plt.xlabel("x")
plt.ylabel("y")
plt.show()
plt.close()
